core/modules/layout/layout.module

```diff
@@ -1471,6 +1500,113 @@ function layout_get_context_info($plugin_name) {
   return isset($info[$plugin_name]) ? $info[$plugin_name] : FALSE;
 }
 
+
+/**
+ * Fetch all relevant relationships.
+ *
+ * Relevant relationships are any relationship that can be created based upon
+ * the list of existing contexts. For example, the 'node author' relationship
+ * is relevant if there is a 'node' context, but makes no sense if there is
+ * not one.
+ *
+ * @param $contexts
+ *   An array of contexts used to figure out which relationships are relevant.
+ *
+ * @return
+ *   An array of relationship keys that are relevant for the given set of
+ *   contexts.
+ */
+function layout_relationships_get_relevant_info($contexts) {
+  $relevant = array();
+  $relationships = _layout_get_all_info('layout_relationship');
+
+  foreach ($contexts as $context) {
+    // Go through each relationship
+    foreach ($relationships as $rid => $relationship) {
+      $handler = layout_create_handler('layout_relationship', $rid);
+      if ($children = $handler->getChildren($rid)) {
+        foreach ($children as $child_rid => $child) {
+          if ($child['context'] == $context->plugin) {
+            $relevant[$child_rid] = $child;
+          }
+        }
+      }
+      else {
+        if ($relationship['context'] == $context->plugin) {
+          $relevant[$rid . ':relationship'] = $relationship;
+        }
+      }
+    }
+  }
+
+  return $relevant;
+}
+
+
+/**
+ * @param $relationship
+ *   A relationship object.
+ *
+ * @param $source_context
+ *   The context this relationship is based upon.
+ *
+ * @return
+ *   A context object if one can be loaded.
+ */
+function layout_get_context_from_relationship($relationship, $source_context) {
+  $new_context = clone($source_context);
+  $context = $relationship->getContext($new_context);
+  if ($context) {
+    return $context;
+  }
+  return array();
+}
+
+/**
+ * Fetch all active relationships
+ *
+ * @param $relationships
+ *   An keyed array of relationship objects.
+ *
+ * @param $contexts
+ *   A keyed array of contexts used to figure out which relationships
+ *   are relevant. New contexts will be added to this.
+ */
+function layout_get_context_from_relationships($relationships, &$contexts) {
+  $contexts_from_relationships_list  = &backdrop_static('layout_get_context_from_relationships', array());
```

Comment:
Looking at the way this is used, I think this should be a new method on the `Layout` class rather than a separate function. Everywhere this is used looks like this currently:

```
    $contexts = $layout->getContexts();

    // Get contexts from relationships
    layout_get_context_from_relationships($layout->relationships, $contexts);
```

And `$contexts` is modified by reference.

I think instead these calls should be assembled like this:

```
    $contexts = $layout->getContexts();
    $contexts += $layout->getContextsFromRelationships()
```

Or possibly, `$layout->getContexts()` just loads the relationships and gets the contexts from relationships all the time, or takes a parameter for `$include_relationships` that defaults to `FALSE`. Then it'd be simplified down to:


```
    $contexts = $layout->getContexts(TRUE);
```

In any case, `getContextsFromRelationships()` should still be its own method, so relationship contexts can be selected individually. We might just call it from within `getContexts()` if the `$include_relationships` flag is `TRUE`.

Reply:
I incorporated a change similar what I describe here in a PR on top of this PR for @docwilmot to review at https://github.com/docwilmot/backdrop/pull/6.